Fix RA Parking/Homing and Slew Limits

ClutchplateDude opened the pull request:
- Switched GoHome functionality to use stepper positions instead of celestial coordinates.
- Fixed a bug in RA Autohoming
- Cleaned up logging some more

CTetford committed:
Merge pull request #7 from OpenAstroTech/develop

Upstream

CTetford committed:
Merge pull request #8 from OpenAstroTech/develop

Pull upstream

CTetford committed:
Merge pull request #9 from OpenAstroTech/develop

Update

ClutchplateDude committed:
V1.10.6 - Updates

- Switched GoHome functionality to use stepper positions instead of celestial coordinates.
- Fixed a bug in RA Autohoming
- Cleaned up logging some more

ClutchplateDude committed:
Merge branch 'develop' into home-by-steps

ClutchplateDude committed:
Version update

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1326,6 +1326,75 @@ void Mount::startSlewingToTarget()
+// Takes any sync operations that have happened and tracking into account.
+void Mount::startSlewingToHome()
+{
+    if (isGuiding())
```
Can we fold this conditional into `stopGuiding()`? Since you'd always use them together

ClutchplateDude replied:
Yup

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1610,9 +1678,7 @@ void Mount::park()
 void Mount::goHome()
 {
-    stopGuiding();
-    setTargetToHome();
```
We can remove `setTargetToHome` now 😄

ClutchplateDude replied:
Well, we can remove `goHome()`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -2366,6 +2432,38 @@ long Mount::getHomingOffset(StepperAxis axis)
 }
 
 #if USE_HALL_SENSOR_RA_AUTOHOME == 1
+String Mount::getHomingState(HomingState state) const
```
Could replace with
```C++
String Mount::getHomingState(HomingState state) const
{
    MappedDict<HomingState, String>::DictEntry_t lookupTable[] = {
        {HOMING_MOVE_OFF, F("MOVE_OFF")},
        {HOMING_MOVING_OFF, F("MOVING_OFF")},
        {HOMING_STOP_AT_TIME, F("STOP_AT_TIME")},
        {HOMING_WAIT_FOR_STOP, F("WAIT_FOR_STOP")},
        {HOMING_START_FIND_START, F("START_FIND_START")},
        {HOMING_FINDING_START, F("FINDING_START")},
        {HOMING_FINDING_START_REVERSE, F("FINDING_START_REVERSE")},
        {HOMING_FINDING_END, F("FINDING_END")},
        {HOMING_RANGE_FOUND, F("RANGE_FOUND")},
        {HOMING_FAILED, F("FAILED")},
        {HOMING_SUCCESSFUL, F("SUCCESSFUL")},
        {HOMING_NOT_ACTIVE, F("NOT_ACTIVE")},
    };
    auto strLookup = MappedDict<HomingState, String>(lookupTable, ARRAY_SIZE(lookupTable));
    String rtnStr;
    if (strLookup.tryGet(state, &rtnStr))
    {
        return rtnStr;
    }
    return F("WTF_STATE");
}
```
It's a bit more terse for lookups like this

ClutchplateDude replied:
I like it.

ClutchplateDude committed:
PR Updates

ClutchplateDude committed:
Code format update

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1402,7 +1397,10 @@ void Mount::startSlewingToHome()
 void Mount::stopGuiding()
 {
-    stopGuiding(true, true);
+    if (isGuiding())
```
The conditional should be in `stopGuiding(bool ra, bool dec)`, not `stopGuiding()`
If you're concerned about indentation change, the early return pattern works nicely here:
```C++
if (!isGuiding())
{
    return;
}
```

ClutchplateDude committed:
PR Feedback Updates

CTetford commented:
@ClutchplateDude have you tested the following scenarios: 
- tracking for 12 hrs and attempting to home?
- tracking for 12 hrs and attempting park (I assume this should be effectively the same thing)
- tracking for 12 hrs and slewing to the current target RA value of what the home is at the time of slew (judging using Stellarium), but not by using the "home" button 
- tracking for 12 hrs and slewing back to a new target in the east

We need to be thorough that it actually fixes all the scenarios that have been reported with slewing incorrectly. It looks like this is only a partial fix so far

ClutchplateDude commented:
I did not test any 12h scenarios. I did a bunch of sync/slew/sync scenarios and some shorter tracking.
What makes you say this is a partial fix? 
Oh, because this only fixes homing after long tracking? Yeah, we still should take your fixes for the target calculation.
I can integrate them in this PR if you like.

CTetford commented:
It looks like the revision only affects returning home, not slewing overall. I think the root of the problem is still `calculateRAandDECSteppers`

ClutchplateDude commented:
> It looks like the revision only affects returning home, not slewing overall. 
Correct, we still need your fix like I said

ClutchplateDude committed:
Merged Camerons limit checking code

CTetford commented:
I'll run some tests on this branch during the day this week. This kind of change needs some thorough testing before we merge, given the implications of it goes wrong

ClutchplateDude commented:
Sounds good. I am also running some, currently with de-energized steppers, but will try with steppers too

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.hpp

CTetford committed:
Update Mount.cpp

ClutchplateDude committed:
WIP - Updates

- Added Meade command so set tracking steps. Will probably remove once homing/sync/slew/track issue is resolved.
- Fixed bug in homing that did not reset DEC stepper to zero (causes problems if a sync previously happened).
- Updated Polaris HA
- Added some more verbose logging.

ClutchplateDude committed:
Code format fix

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.cpp

CTetford committed:
Merge branch 'develop' into home-by-steps

CTetford committed:
Merge branch 'develop' into home-by-steps

CTetford committed:
Update Version.h

CTetford committed:
Update Mount.cpp

CTetford committed:
Merge branch 'develop' into home-by-steps

CTetford committed:
Clang-format

CTetford committed:
Update test_sidereal.h

ClutchplateDude committed:
Adjust zeroRA when syncing for RA.

CTetford committed:
Update Mount.cpp

CTetford committed:
Add zeroPosDEC

CTetford committed:
Update Mount.cpp

CTetford committed:
Fix DEC limitation

CTetford committed:
Disable Tracking when RA Limit Hit

Add function to check current RA position/tracking and disable tracking if the limit of the ring is hit.

CTetford committed:
Configurable DEC Axis Limits

Add config items for DEC limitations

CTetford committed:
Fixes

CTetford committed:
Update Changelog.md

ClutchplateDude committed:
WIP updated LCD code

ClutchplateDude committed:
Revert last change to c71

CTetford committed:
Update Mount.cpp

ClutchplateDude committed:
Updates

Limit checks throttled to every 5s.
Bug fix in RA Hall Homing.
Some logging fixes.

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1179,19 +1191,20 @@ const DayTime Mount::currentRA() const
-          String(siderealDegreesInHour, 5).c_str());
-    LOGV2(DEBUG_MOUNT_VERBOSE, F("[MOUNT]: CurrentRA: RA Steps   : %d"), _stepperRA->currentPosition());
-    LOGV2(DEBUG_MOUNT_VERBOSE, F("[MOUNT]: CurrentRA: POS        : %s"), String(hourPos).c_str());
+    // LOGV4(DEBUG_MOUNT_VERBOSE,
```
Remove commented code

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1179,19 +1191,20 @@ const DayTime Mount::currentRA() const
 
-    bool flipRA = _stepperDEC->currentPosition() < 0;
-    if (flipRA)
+    float degreePos = _stepperDEC->currentPosition() / _stepsPerDECDegree;
```
`const float degreePos = (_stepperDEC->currentPosition() / _stepsPerDECDegree) + _zeroPosDEC;`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1222,6 +1235,8 @@ const Declination Mount::currentDEC() const
     //LOGV2(DEBUG_MOUNT_VERBOSE,F("[MOUNT]: CurrentDEC: DEC Steps  : %d"), _stepperDEC->currentPosition());
     //LOGV2(DEBUG_MOUNT_VERBOSE,F("[MOUNT]: CurrentDEC: POS        : %s"), String(degreePos).c_str());
 
+    degreePos += _zeroPosDEC;
```
Same thing here, but you could even refactor the entire function here:
```C++
const Declination Mount::currentDEC() const
{
    // u-steps / u-steps/deg = deg
    const float degreePos = (_stepperDEC->currentPosition() / _stepsPerDECDegree) + _zeroPosDEC;

    if (NORTHERN_HEMISPHERE ? degreePos > 0 : degreePos < 0)
    {
        return -degreePos;
    }
    return degreePos;
}
```

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1242,25 +1257,33 @@ const Declination Mount::currentDEC() const
+    float raAdjust = ra.getTotalHours() - currentRA().getTotalHours();
+    while (raAdjust > 12)
+    {
+        raAdjust = raAdjust - 24;
```
`raAdjust -= 24.0f;`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1242,25 +1257,33 @@ const Declination Mount::currentDEC() const
+    }
+    while (raAdjust < -12)
+    {
+        raAdjust = raAdjust + 24;
```
`raAdjust += 24.0f;`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1242,25 +1257,33 @@ const Declination Mount::currentDEC() const
+    _zeroPosRA.addHours(raAdjust);
+
+    // Adjust the home DEC position by the delta between the sync'd target and current position.
+    float decAdjust = dec.getTotalDegrees() - currentDEC().getTotalDegrees();
```
`const float decAdjust`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1242,25 +1257,33 @@ const Declination Mount::currentDEC() const
 
     long targetRAPosition, targetDECPosition;
-    LOGV3(DEBUG_MOUNT, F("[MOUNT]: Sync Position to RA: %s and DEC: %s"), _targetRA.ToString(), _targetDEC.ToString());
+    LOGV3(DEBUG_MOUNT, F("[MOUNT]: Sync Position to RA: %s and DEC: %s"), ra.ToString(), _targetDEC.ToString());
```
Shouldn't this log be both `ra, dec` or `_targetRA, _targetDEC` and not one of either?

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1328,46 +1348,110 @@ void Mount::startSlewingToTarget()
+
+    _slewingToHome = true;
+    // Take tracking into account
+    long trackingOffset = _stepperTRK->currentPosition() * RA_SLEW_MICROSTEPPING / RA_TRACKING_MICROSTEPPING;
```
`const long trackingOffset`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1328,46 +1348,110 @@ void Mount::startSlewingToTarget()
+
+    // Take any syncs that have happened into account
+    long targetRAPosition  = -_homeOffsetRA;
+    long targetDECPosition = -_homeOffsetDEC;
```
Technically you can remove this variable (and replace with `-_homeOffsetDEC`) but if the name is useful it should at least be `const`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1328,46 +1348,110 @@ void Mount::startSlewingToTarget()
+    moveSteppersTo(targetRAPosition, targetDECPosition);  // u-steps (in slew mode)
+
+    _mountStatus |= STATUS_SLEWING | STATUS_SLEWING_TO_TARGET;
+    _totalDECMove = 1.0f * _stepperDEC->distanceToGo();
```
```C++
_totalDECMove = static_cast<float>(_stepperDEC->distanceToGo());
_totalRAMove  = static_cast<float>(_stepperRA->distanceToGo());
```
instead of the `1.0f *`

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1328,46 +1348,110 @@ void Mount::startSlewingToTarget()
 {
-    // Stop RA guide first, since it's just a speed change back to tracking speed
-    if (ra && (_mountStatus & STATUS_GUIDE_PULSE_RA))
+    if (isGuiding())
```
You can use an early return pattern to clean up the scope
```C++
if (!isGuiding())
{
    return;
}
```

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -3163,41 +3270,69 @@ void Mount::calculateStepperPositions(float raCoord, float decCoord, long &raPos
+        homeTargetDeltaRA = homeTargetDeltaRA + 24;
+    }
+
     // Map [0 to 24] range to [-12 to +12] range
```
Seems like we keep doing a lot of this, maybe a future improvement could abstract this kind of operation :smiley:

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -3740,3 +3884,33 @@ void Mount::testUART_vactual(TMC2209Stepper *driver, int _speed, int _duration)
+void Mount::checkRALimit()
+{
+    // Check tracking limits every 5 seconds
+    if (millis() - _lastTRKCheck > 5000)
```
Same early return pattern could be used here

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -3740,3 +3884,33 @@ void Mount::testUART_vactual(TMC2209Stepper *driver, int _speed, int _duration)
+    // Check tracking limits every 5 seconds
+    if (millis() - _lastTRKCheck > 5000)
+    {
+        float trackedHours       = (_stepperTRK->currentPosition() / _trackingSpeed) / 3600.0F;  // steps / steps/s / 3600 = hours
```
`trackedHours`, `homeRA`, and `RALimit` can all be `const` :smile: 
(IMO `const float RALimit` instead of `float const RALimit` is the common pattern)

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -3740,3 +3884,33 @@ void Mount::testUART_vactual(TMC2209Stepper *driver, int _speed, int _duration)
+
+        LOGV2(DEBUG_MOUNT_VERBOSE, F("[MOUNT]: checkRALimit: homeRAdelta: %f"), homeCurrentDeltaRA);
+        while (homeCurrentDeltaRA > 12)
+            homeCurrentDeltaRA = homeCurrentDeltaRA - 24;
```
`homeCurrentDeltaRA -= 24.0f` helps to show intent here by making the statement a bit less verbose

julianneswinoga commented on test/test_embedded/main.cpp:
```diff
@@ -0,0 +1,24 @@
+
+#include "test_sidereal.h"
+
+// void setup()
```
Remove commented code

julianneswinoga commented on test/test_embedded/main.cpp:
```diff
@@ -0,0 +1,24 @@
+{
+    UNITY_BEGIN();
+
+    test::sidereal::run();
```
Just curious, are these tests actually running on the target? @andre-stefanov had some trouble running embedded tests

julianneswinoga replied:
Reminder to remove these files too

julianneswinoga reviewed (changes requested):
Mostly minor things, looks awesome overall!

CTetford committed:
Updates

CTetford committed:
Merge branch 'develop' into home-by-steps

CTetford committed:
Update Version.h

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.cpp

ClutchplateDude committed:
WIP Declination change

ClutchplateDude committed:
WIP - allow full DEC range inputs

CTetford committed:
Add separate debug serial

CTetford committed:
Update Declination.cpp

ClutchplateDude committed:
Fix some abs() calls causing overflows

CTetford committed:
Fix abs()

CTetford committed:
Merge branch 'home-by-steps' of https://github.com/OpenAstroTech/OpenAstroTracker-Firmware into home-by-steps

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.cpp

CTetford committed:
Update Mount.cpp

CTetford committed:
Configurable Debug Serial Port

CTetford committed:
Update Changelog.md

ClutchplateDude committed:
Merge latest

ClutchplateDude committed:
Added CALC logging

ClutchplateDude committed:
Format change

ClutchplateDude committed:
Removed test code

julianneswinoga commented on ConfigurationValidation.hpp:
```diff
@@ -282,6 +282,13 @@
 #endif
 
+// Debugging
+#if DEBUG_SEPARATE_SERIAL == 1
```
Why do we need `DEBUG_SEPARATE_SERIAL`? `DEBUG_SERIAL_PORT` should be enough

ClutchplateDude replied:
Removed.

CTetford replied:
See my other comment. I've changed the implementation but kept this flag as an option. If someone wants to specify a non-default debug port directly in their config they can, but the DEBUG_SEPARATE_SERIAL flag is still needed.

julianneswinoga commented on boards/AVR_MKS_GEN_L_V1/pins_MKS_GEN_L_V1.h:
```diff
@@ -137,3 +137,8 @@
+
+//Serial port for external debugging
+#ifndef DEBUG_SERIAL_PORT
+    #define DEBUG_SERIAL_PORT Serial3  //D14 (Y-MIN) - TX3 and D15 (Y-MAX) - RX3
```
I think the defaults should still be USB serial, debugging for users would get a lot harder if we asked them to hook up an extra serial line

CTetford replied:
I've revised the implementation. DEBUG_SEPARATE_SERIAL is still a boolean option, but I've removed some redundant code and the pins files default to the main serial channel for the debug output.

julianneswinoga commented on Configuration_adv.hpp:
```diff
@@ -559,6 +559,14 @@
+#endif
+
+#ifndef DEBUG_SERIAL_BAUDRATE
+    #define DEBUG_SERIAL_BAUDRATE 57600
```
Why `57600`? `115200` is a "standard" mid-speed serial connection

julianneswinoga commented on src/Declination.cpp:
```diff
@@ -93,20 +102,24 @@ Declination Declination::ParseFromMeade(String const &s)
 {
-    seconds += (NORTHERN_HEMISPHERE ? -(arcSecondsPerHemisphere / 2) : (arcSecondsPerHemisphere / 2));
-    return Declination(1.0 * seconds / 3600.0);
+    if (NORTHERN_HEMISPHERE)
```
The float conversion should be done upfront to avoid a precision loss:
```C++
Declination Declination::FromSeconds(long seconds)
{
    const auto secondsFloat                 = static_cast<float>(seconds);
    const auto arcSecondsPerHemisphereFloat = static_cast<float>(arcSecondsPerHemisphere);
#if NORTHERN_HEMISPHERE == 1
    return Declination(((arcSecondsPerHemisphereFloat / 2.0f) - secondsFloat) / 3600.0f);
#else
    return Declination(((-arcSecondsPerHemisphereFloat / 2.0f) + secondsFloat) / 3600.0f);
#endif
}
```

ClutchplateDude replied:
Good catch

julianneswinoga commented on src/Mount.cpp:
```diff
@@ -1220,13 +1220,14 @@ const Declination Mount::currentDEC() const
-    {
-        return -degreePos;
-    }
+    // if (NORTHERN_HEMISPHERE ? degreePos > 0 : degreePos < 0)
```
Remove commented code

ClutchplateDude replied:
Done.

CTetford committed:
Revise DEBUG_SERIAL_PORT

CTetford committed:
Merge branch 'develop' into home-by-steps

CTetford committed:
Update Version.h

ClutchplateDude committed:
PR feedback changes

Bumped debug serial to 115200
cast floats earlier in dec
removed commented code

ClutchplateDude committed:
Merge branch 'develop' into home-by-steps

Version bump

ClutchplateDude committed:
Version bump in changelog

ClutchplateDude committed:
Fix copy/paste error in serial

ClutchplateDude committed:
Format update

julianneswinoga commented on src/f_serial.hpp:
```diff
@@ -43,8 +43,10 @@ void processSerialData()
+                // When not debugging, print the result to the serial port .
+                // When debugging, only print the result to Serial if we're on seperate ports.
+    #if (DEBUG_LEVEL == DEBUG_NONE) || (DEBUG_SEPARATE_SERIAL == 1)
+                Serial.print("1");
```
The implementation has changed here I think, I'm not too sure what this code is supposed to be doing but you've changed the serial response from `0x31` (`'1'`) to `0x31 0x0` (`'1'` with null termination)

ClutchplateDude replied:
Oops, good catch.

ClutchplateDude committed:
Revert incorrect handshake reply

ClutchplateDude committed:
Removed embedded test files.
